Update .README for take home test (#6)
Accurx/reliable-downloader: README.md

# 👩‍⚕️ Reliable Downloader Exercise 🖥️

[Find out more about Accurx](https://www.accurx.com/careers)

Thank you for investing your time in our take-home exercise.

We've based this exercise on a real problem we've had to solve to be able to release our desktop software, which allows clinicians to communicate with patients over text, video, questionnaires, and other methods to hundreds of thousands of users multiple times a week.

### Why have we set this exercise?
So we get an insight into how you:
* Understand and  analyse requirements to solve real user problems
* Utilise language features, structure code and tests to verify your solution meets user requirements. 

### What is expected of you?

> 1. Please complete the task detailed below👇

Reviewer: This feels odd to have a numbered point, but I get the intent and don't have a great alternative suggestion.

Author: Updated to use bold type instead

2. Please answer the following questions in `questions.md`
   - How did you approach solving the problem?
   - How did you verify your solution works correctly?
   - How long did you spend on the exercise?
   - What would you add if you had more time and how?
4. When you're finished please:
   - Download your solution including your completed `questions.md` file (in GitHub, at the route of your repository, click Code -> Download Zip)
   - Submit your zipped solution, using the link in your invite email.

_Please feel free to add any feedback you have on this exercise in the submission `feedback.md`_

## Task
### Context
The component that clinicians use for  downloading updates needs to be reliable in the unreliable network conditions they work with; often facing intermittent internet disconnection and slow internet speeds.

We would like you to update the skeleton project in this repository to provide a way for clinicians to download our software given multiple network conditions. Details can be found in the requirements section below. 

### Requirements

1. The installer downloads in the following scenarios:

>    - When internet disconnections occur (we use 2 minutes as a disconnection time benchmark for this)

Reviewer: Is there a reason we're avoiding the word "timeout" here?

Author: Agree this is pretty ambiguous language. I've replaced with the following 

- The download is successful if the internet connection is lost for up to two minutes
- The downloader should timeout if the internet connection is lost for more than two minutes

   - Partial downloading. So that the download doesn't need to start from scratch every time if the CDN supports this
   - Downloading the file in one go, if the CDN does not support partial downloading
2. The system can recover from failures and not exit until the file has been successfully downloaded
3. The file is deleted if the integrity check fails after downloading. You can use the Content-MD5 for this: https://www.oreilly.com/library/view/http-the-definitive/1565925092/re17.html
4. Progress is reported to the user throughout the download
5. The user can cancel the download

### What does a successful submission look like?
A high-quality submission will have the following:
- Meet the requirements above

> - Appropriate use of language features to solve the problem in a neat and simple way

Reviewer: ```suggestion
- Appropriate use of language features to solve the problem in a simple way
```
I don't think neat really adds anything here

> - Well structured code

Author: What does well-structured code look like? Can we be more specific about the things we're looking for here?

Reviewer 2: I think we're looking for code that is easy to read and reason about. Typically using software design principles like SOLID.

Author: Thanks Jaf! I've made an update based on this - really helpful!

> - Code that is easy to read and reason about. Typically using software design principles like SOLID.

Reviewer: ```suggestion
- Code that is easy to read and reason about.
```
I have _thoughts_ and _feelings_ on SOLID and I'm not sure that the second sentence here really adds anything.

Reviewer: Oh no, I see that @JafarMiles-AccuRx and I have just given opposite suggestions 😿 , although at least we agree on the first part.

Author: I think that the cut-down version is OK. I think Jafs explanation of what is meant here was helpful for me but would expect an engineer PE3 onwards to understand this. Thanks both!

- Unit tests

_If you feel that modifying the skeleton project would create a better solution to these requirements, we encourage you to do so. We currently support submissions in either .NET (C#) or Java and have included skeleton projects for both._

## Tips

- Take the time to read through the task and description. There's guidance in there that can be helpful to approaching the problem.

> - The code doesn't need to be perfect but it needs to be readable.

Reviewer 3: If we're modifying the README - i'd personally like to remove this line
(let the candidate determine what quality code they're happy to submit)

Author: Good suggestion Jimbo - thanks! removed

- Try writing down some example inputs and outputs on paper.
- Try a brute force approach and then optimise the code.
- Add some comments to your code if you think it will be helpful to share your thought process to someone assessing it.
- You can throttle your internet connection using NetLimiter or similar.
- You can simulate internet disconnections by disconnecting wifi/ethernet.
- Different behaviours occur after following different periods of disconnection, two seconds and two minutes are sweet spots for exercising key failure modes.

> - The use of NuGet packages/third-party libraries is fine, however, do bear in mind that overusing these limits your ability to show off!

Reviewer: ```suggestion
- The use of third-party libraries is fine, however, do bear in mind that overusing these may limit your ability to demonstrate your skills
```
NuGet is .NET specific, so I'd suggest that we just say third-party libraries unless we want to move this into language specific READMEs.
